shared/p2p/service_test.go

```diff
@@ -132,3 +161,108 @@ func testSubscribe(ctx context.Context, t *testing.T, s Server, gsub *floodsub.P
 		t.Error("Context timed out before a message was received!")
 	}
 }
+
+func TestRegisterTopic_WithoutAdapters(t *testing.T) {
+	s, err := NewServer()
+	if err != nil {
+		t.Fatalf("Failed to create new server: %v", err)
+	}
+	topic := "test_topic"
+	testMessage := testpb.TestMessage{Foo: "bar"}
+
+	s.RegisterTopic(topic, testMessage)
+
+	ch := make(chan Message)
+	sub := s.Subscribe(testMessage, ch)
+	defer sub.Unsubscribe()
+
+	wait := make(chan struct{})
+	go (func() {
+		defer close(wait)
+		msg := <-ch
+		_ = msg
+	})()
+
+	if err := simulateIncomingMessage(t, s, topic, []byte{}); err != nil {
+		t.Errorf("Failed to send to topic %s", topic)
+	}
+
+	select {
+	case <-wait:
+		return // OK
+	case <-time.After(1 * time.Second):
+		t.Fatal("TestMessage not received within 1 seconds")
+	}
+}
+
+func TestRegisterTopic_WithAdapers(t *testing.T) {
+	s, err := NewServer()
+	if err != nil {
+		t.Fatalf("Failed to create new server: %v", err)
+	}
+	topic := "test_topic"
+	testMessage := testpb.TestMessage{Foo: "bar"}
+
+	i := 0
+	var testAdapter Adapter = func(next Handler) Handler {
+		return func(ctx context.Context, msg Message) {
+			i++
+			next(ctx, msg)
+		}
+	}
+
+	adapters := []Adapter{
+		testAdapter,
+		testAdapter,
+		testAdapter,
+		testAdapter,
+		testAdapter,
+	}
+
+	s.RegisterTopic(topic, testMessage, adapters...)
+
+	ch := make(chan Message)
+	sub := s.Subscribe(testMessage, ch)
+	defer sub.Unsubscribe()
+
+	wait := make(chan struct{})
+	go (func() {
+		defer close(wait)
+		msg := <-ch
+		_ = msg
+	})()
+
+	if err := simulateIncomingMessage(t, s, topic, []byte{}); err != nil {
+		t.Errorf("Failed to send to topic %s", topic)
+	}
+
+	select {
+	case <-wait:
+		if i != 5 {
+			t.Errorf("Expected testAdapter to increment i to 5, but was %d", i)
+		}
+		return // OK
+	case <-time.After(1 * time.Second):
+		t.Fatal("TestMessage not received within 1 seconds")
+	}
+}
+
+func simulateIncomingMessage(t *testing.T, s *Server, topic string, b []byte) error {
+	ctx := context.Background()
+	h := bhost.NewBlankHost(swarmt.GenSwarm(t, ctx))
+
+	gsub, err := floodsub.NewFloodSub(ctx, h)
+	if err != nil {
+		return err
+	}
+
+	pinfo := h.Peerstore().PeerInfo(h.ID())
+	if err = s.host.Connect(ctx, pinfo); err != nil {
+		return err
+	}
+
+	// Short timeout to allow libp2p to handle peer connection.
+	time.Sleep(time.Millisecond * 10)
```

Comment:
The description for `BasicHost.Connect` is:
```
// Connect ensures there is a connection between this host and the peer with
// given peer.ID. If there is not an active connection, Connect will issue a
// h.Network.Dial, and block until a connection is open, or an error is returned.
// Connect will absorb the addresses in pi into its internal peerstore.
// It will also resolve any /dns4, /dns6, and /dnsaddr addresses.
```

which suggests that this timeout shouldn't be necessary. Do you know why it's required? This is a potential source of flakiness, we should definitely try and avoid an arbitrary timeout.

Reply:
Not really. This pattern is all over the floodsub tests: https://github.com/libp2p/go-floodsub/blob/HEAD/floodsub_test.go

And it fails without it.